Add links to artists and render homepage link

philnash opened the pull request:
- Fetches links from MusicBrainz by the artist mbid
- Renders a homepage link under albums if the artist has an "official homepage" link

TODO:

- Either add an `ArtistServiceCheck` model or generalise the `AlbumServiceCheck` model with a polymorphic association
- Only update links for artists without a recent check

---

The first thing I thought of when trying out Support Act was that I'd like to buy records from local record stores or the artist themselves. Now, there's no API for whether your closest record store has the album you want (could this be wrangled out of [Discogs](https://www.discogs.com/developers) somehow? Possibly, but with great difficulty). I did hope I could at least link to the artist's homepage. Turns out MusicBrainz does have that information, so I mimicked the Album update from MusicBrainz for artists, giving them a JSON links attribute too (you never know what other links might be useful, following a band on a social media can present further opportunities to support them, for example). This now renders the artist homepage if one is found in MusicBrainz data, if an mbid is available for the artist.

As per the TODO above, I wondered first whether this was a feature you'd like and if so what option you would prefer? It's been a while since I messed with a database and a polymorphic association, but that likely seems the best approach to me.

What do you think?

pat commented:
Thanks Phil :) I'm definitely on board with this PR - and I think `AlbumServiceCheck` should be generalised to be just `ServiceCheck`, with a polymorphic association to cover both artists and albums.

I don't suppose you've got a screenshot on hand for how the artist links look? I'm thinking the album elements are getting rather crowded now… it'd be nice to compact things down a bit, but not quite sure what that should look like 🤔

philnash commented:
It is getting a bit unwieldy. I'll have a think about making that a bit more compact!
![Support Act 2021-01-10 14-55-35](https://user-images.githubusercontent.com/31462/104114039-ff90c200-5353-11eb-9496-07c9b2036d8b.png)

philnash commented:
Ok, updated the `AlbumServiceCheck` to be a `ServiceCheck` and applied it to `Artist`s as well.

Also, changed the Artist homepage link to their name rather than adding another button. And added that to the emails too, which I forgot the first time around.

I feel good about this now. (Might be the most complicated Rails application change I've done for a while!)

What do you think?

pat commented:
All looks great! I'll merge it in and we'll see how it goes :)

philnash commented:
🍾

pat commented:
Had to make a couple of small tweaks (4f6bb6719d8023054d6abfcf9cd72887a5aca7c1 and f8dcbf690a30bfa6376e6b3f99097fc282f42dfc) but it's now on production, the task is running, and there are artist links gradually appearing 🙌🏻 (currently up to artist 500 of ~> 2700)

philnash commented:
Oh, didn't realise the foreign key thing would cause problems. And I honestly thought I'd added that enumerator, more than once even 🤦‍♂️. Thanks for fixing it up!
